### PR TITLE
[DIA-2360] Fix message showing multiple times on reconsent flow

### PR DESCRIPTION
**Pull request by andresilveirah**
@bohdan-go-wombat @Nevazhnovu admittedly this PR is doing more than it's supposed to, but all the changes were done in the process of getting to the bottom of the issue. 

The issue [DIA-2360](https://sourcepoint.atlassian.net/browse/DIA-2360) was caused by having "mirroring values" for `gdprDateCreated` so, strictly speaking only changes referring to `gdprDateCreated` are relevant to this PR. 

All other changes are steps towards a cleaner/clearer SDK. 

[DIA-2360]: https://sourcepoint.atlassian.net/browse/DIA-2360?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: samples/app/src/androidTest/java/com/sourcepoint/app/v6/SpClientMock.kt [hunk 1]
```diff
@@ -1,32 +0,0 @@
...
-
-import android.view.View
-import com.sourcepoint.cmplibrary.NativeMessageController
-import com.sourcepoint.cmplibrary.SpClient
-import com.sourcepoint.cmplibrary.core.nativemessage.MessageStructure
-import com.sourcepoint.cmplibrary.model.ConsentAction
-import com.sourcepoint.cmplibrary.model.exposed.SPConsents
-import org.json.JSONObject
-
-class SpClientMock : SpClient {
```
Comment (bohdan-go-wombat) [inline samples/app/src/androidTest/java/com/sourcepoint/app/v6/SpClientMock.kt:11]: you decided to remove this class to mock it explicitly in each test suite, right?

Comment (andresilveirah) [reply samples/app/src/androidTest/java/com/sourcepoint/app/v6/SpClientMock.kt:11]: @bohdan-go-wombat this was used only in one test case. I refactor that test case to use a `@Mockk` instance instead.

File: cmplibrary/src/main/java/com/sourcepoint/cmplibrary/data/local/DataStorageGdpr.kt [hunk 2]
```diff
@@ -332,14 +301,4 @@ private class DataStorageGdprImpl(context: Context) : DataStorageGdpr {
             .forEach { entry -> spEditor.remove(entry.key) }
         spEditor.apply()
     }
-
-    private fun fail(param: String): Nothing = throw RuntimeException("$param not fund in local storage.")
-}
-
-internal fun DataStorageGdpr.getGDPRConsent(): Either<GDPRConsentInternal> = check {
```
Comment (bohdan-go-wombat) [inline cmplibrary/src/main/java/com/sourcepoint/cmplibrary/data/local/DataStorageGdpr.kt:339]: why did you delete this part? was it unused?

File: cmplibrary/src/main/java/com/sourcepoint/cmplibrary/data/ServiceImpl.kt [hunk 3]
```diff
@@ -48,13 +48,13 @@ internal fun Service.Companion.create(
  * Implementation os the [Service] interface
  */
 private class ServiceImpl(
-    private val nc: NetworkClient,
+    private val networkClient: NetworkClient,
```
Comment (bohdan-go-wombat) [inline cmplibrary/src/main/java/com/sourcepoint/cmplibrary/data/ServiceImpl.kt:51]: big PLUS for that!! 👍

File: samples/app/src/androidTest/java/com/sourcepoint/app/v6/MainActivityKotlinNativeMessage.kt [hunk 4]
```diff
@@ -32,8 +32,11 @@ import org.koin.core.context.loadKoinModules
 class MainActivityNativeMessTest {
 
     lateinit var scenario: ActivityScenario<NativeMessageActivity>
-    private val appContext by lazy { InstrumentationRegistry.getInstrumentation().targetContext }
 
+    @Before
+    fun cleanLocalStorage() {
```
[the remark this answers is not in the document]
Comment (andresilveirah) [reply samples/app/src/androidTest/java/com/sourcepoint/app/v6/MainActivityKotlinNativeMessage.kt:37]: Got it. I'm aware of the `setup` / `teardown` practice, but methods annotated with `@Before` run before _every_ test case, so in a sense they are not really classic cases of test setup/teardown.

But we can agree in having always a single `@Before` method and name it `setup` or having several `@Before` for more verbosity on what each method is doing.

File: ui-test-util/src/main/java/com/example/uitestutil/TestUtils.kt [hunk 5]
```diff
@@ -9,15 +9,15 @@ import kotlin.jvm.Throws
 
 @Throws(Throwable::class)
 suspend fun wr(
-    delay: Long = 200,
-    times: Int = 30,
+    delayInMs: Long = 10,
+    times: Int = 2000,
```
[the remark this answers is not in the document]
Comment (andresilveirah) [reply ui-test-util/src/main/java/com/example/uitestutil/TestUtils.kt:13]: If I understand right, the `wr` method is applying a "pooling" mechanism. That way, it's more efficient to let the machine _pool_ several times and check for the task's result instead of a few times with a longer delay. (also that seems to fix some of the UI tests flakiness).

File: cmplibrary/src/main/java/com/sourcepoint/cmplibrary/data/local/DataStorageGdpr.kt (hunk 2 above)
Comment (andresilveirah) [reply cmplibrary/src/main/java/com/sourcepoint/cmplibrary/data/local/DataStorageGdpr.kt:339]: Yes. There's A LOT of unused code... I think we need a separate task only to hunt those down and delete them.

File: cmplibrary/src/main/java/com/sourcepoint/cmplibrary/data/ServiceImpl.kt [hunk 6]
```diff
@@ -233,25 +226,21 @@ private class ServiceImpl(
...
-                    .executeOnRight {
-                        it.gdpr?.uuid?.let { u ->
-                            campaignManager.gdprUuid = u
-                            campaignManager.gdprConsentStatus = campaignManager.gdprConsentStatus?.copy(
-                                uuid = u
-                            )
-                        }
+                    .executeOnRight { pvDataResponse ->
+                        campaignManager.gdprConsentStatus = campaignManager.gdprConsentStatus?.copy(
+                            uuid = pvDataResponse.gdpr?.uuid
```
Comment (bohdan-go-wombat) [inline cmplibrary/src/main/java/com/sourcepoint/cmplibrary/data/ServiceImpl.kt:242]: that could save null as the `uuid`, and the previous code did not change the value of the `uuid` if the `it.gdpr?.uuid` returned `null`

do you think we are ok with refreshing the value of `uuid` with the `null` in this case?

File: cmplibrary/src/main/java/com/sourcepoint/cmplibrary/data/ServiceImpl.kt [hunk 7]
```diff
@@ -267,25 +256,21 @@ private class ServiceImpl(
...
-                    .executeOnRight {
-                        it.ccpa?.uuid?.let { u ->
-                            campaignManager.ccpaUuid = u
-                            campaignManager.ccpaConsentStatus = campaignManager.ccpaConsentStatus?.copy(
-                                uuid = u
-                            )
-                        }
+                    .executeOnRight { pvDataResponse ->
+                        campaignManager.ccpaConsentStatus = campaignManager.ccpaConsentStatus?.copy(
+                            uuid = pvDataResponse.ccpa?.uuid
```
Comment (bohdan-go-wombat) [inline cmplibrary/src/main/java/com/sourcepoint/cmplibrary/data/ServiceImpl.kt:272]: same here

File: cmplibrary/src/main/java/com/sourcepoint/cmplibrary/data/ServiceImpl.kt [hunk 8]
```diff
@@ -334,18 +318,16 @@ private class ServiceImpl(
                 metadataArg = campaignManager.metaDataResp?.toMetaDataArg()?.copy(ccpa = null),
             )
 
-            getResp = nc.getChoice(gcParam)
-                .executeOnRight { r ->
-                    r.gdpr?.let {
-                        campaignManager.gdprConsentStatus = it.copy(uuid = campaignManager.gdprUuid)
+            getResp = networkClient.getChoice(gcParam)
+                .executeOnRight { response ->
+                    response.gdpr?.let { responseGdpr ->
```
Comment (bohdan-go-wombat) [inline cmplibrary/src/main/java/com/sourcepoint/cmplibrary/data/ServiceImpl.kt:323]: 👍

**Review by bohdan-go-wombat — COMMENTED** [review]
left a bunch of comments

File: cmplibrary/src/main/java/com/sourcepoint/cmplibrary/data/ServiceImpl.kt (hunk 6 above)
Comment (andresilveirah) [reply cmplibrary/src/main/java/com/sourcepoint/cmplibrary/data/ServiceImpl.kt:242]: Yes. The only places that _should_ update uuid are either `pvData` or post choice responses. Even if the `uuid` is null.

File: ui-test-util/src/main/java/com/example/uitestutil/TestUtils.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (bohdan-go-wombat) [reply ui-test-util/src/main/java/com/example/uitestutil/TestUtils.kt:13]: great thing

**Review by bohdan-go-wombat — APPROVED** [review]

**Review by Nevazhnovu — APPROVED** [review]
I wouldn't write it better 👍

File: cmplibrary/src/main/java/com/sourcepoint/cmplibrary/data/local/DataStorageCcpa.kt [hunk 9]
```diff
@@ -53,13 +48,11 @@ internal interface DataStorageCcpa {
         const val KEY_CCPA_CHILD_PM_ID = "sp.ccpa.key.childPmId"
         const val CCPA_CONSENT_RESP = "sp.ccpa.consent.resp"
         const val CCPA_JSON_MESSAGE = "sp.ccpa.json.message"
-        const val CONSENT_CCPA_UUID_KEY = "sp.ccpa.consentUUID"
```
[the remark this answers is not in the document]
Comment (andresilveirah) [reply cmplibrary/src/main/java/com/sourcepoint/cmplibrary/data/local/DataStorageCcpa.kt:56]: No, it should not trigger a re-consent flow. Here I'm removing the key from the local storage, but uuid is still persisted inside the ccpa/gdpr consent objects.

File: samples/metaapp/src/main/java/com/sourcepointmeta/metaapp/util/DefaultProperties.kt [hunk 10]
```diff
@@ -229,13 +229,11 @@ val oldV6Consent = """
...
-          "sp.key.property.id": 16893,
```
[the remark this answers is not in the document]
Comment (andresilveirah) [reply samples/metaapp/src/main/java/com/sourcepointmeta/metaapp/util/DefaultProperties.kt:238]: I did. But it turns out this file is used by the meta app to facilitate adding legacy consent to the shared prefs in order to test the transition from v6 to v7. 

However, it's possible that we'll remove that feature entirely in the future.
